Issue 123 all layers option

juanmatias opened the pull request:
## What?
* "All Layers"feature added
TF base commands can be run as they are from inside a layer dir:
```
leverage tf init
```
...or with `--layers` option from the account dir:
```
leverage tf init --layers us-east-1/terraform-backend,global/sso
```
* the code inside `terraform.py` was refactored for sake of readability.

## Why?
* Fulfill a user request
* Take advantage of the work and refactor code for clarity

## References
n/a

angelofenoglio commented on leverage/modules/terraform.py:
```diff
@@ -140,15 +129,197 @@ def _format(tf, args):
+    for layer in layers:
+        logger.debug(f"Checking for layer {layer}...")
+        # change to current dir and set it in the container
+        os.chdir(layer)
```
Is it really necessary to change location for the running cli? Isn't enough just changing cwd for the container?

juanmatias replied:
Interesting enough... let me test it.

juanmatias replied:
@angelofenoglio we are using the current path from the CLI to check a few things... so we need to change it before running the commands.

juanmatias replied:
it's done, it is working ok

angelofenoglio replied:
Great!

juanmatias commented on leverage/container.py:
```diff
@@ -507,6 +507,35 @@ def _prepare_container(self):
+            return 'config'
+        elif (self.cwd.as_posix().find(self.account_dir.as_posix()) >= 0) and list(self.cwd.glob("*.tf")):
+            return 'layer'
+        elif (self.cwd.as_posix().find(self.account_dir.as_posix()) >= 0) and not list(self.cwd.glob("*.tf")):
+            return 'other under project'
+        else:
+            return 'not a project'
```
I can't think of a scenario (Leverage CLI should fail previously in these scenarios), but I added them since they were just a few lines to cover all the possibilities. (i.e. to handle all possibilities)

angelofenoglio replied:
Ok, I guess that in this case is harmless enough that we can leave it as it is. Thanks for the clarification!

juanmatias commented on leverage/container.py:
```diff
@@ -507,6 +507,35 @@ def _prepare_container(self):
 
         logger.debug(f"[bold cyan]Running with entrypoint:[/bold cyan] {self.entrypoint}")
 
+    def get_location_type(self):
```
It makes sense. Will do.

juanmatias commented on leverage/container.py:
```diff
@@ -507,6 +507,35 @@ def _prepare_container(self):
+        else:
+            return 'not a project'
+
+    def check_for_account_location(self):
```
Functionality assigned to `check_for_account_location` was moved here: https://github.com/binbashar/leverage/blob/cbdb725f8666af41d094d1a7ee4c3350a1fb82e4/leverage/modules/terraform.py#L186
Thus, I'm going to remove the function.

juanmatias commented on leverage/modules/aws.py:
```diff
@@ -174,9 +174,24 @@ def sso(context, cli, args):
+    # - when this cond meets:
+    #   - no account dir
+    #   - no layer dir
+    if (cli.cwd == cli.root_dir or
+        (cli.cwd != cli.account_dir and
+            (
+            (cli.account_dir.parent != cli.account_dir or cli.account_dir.parent.parent != cli.accoununt_dir) and
+            not list(cli.cwd.glob("*.tf"))
+            )
+        )):
```
A lot better!
```
    if not cli.get_location_type() in ['account', 'layer']:
        logger.error("SSO configuration can only be performed at [bold]layer[/bold] level.")
        raise Exit(1)
```

angelofenoglio replied:
Beautiful!

juanmatias commented:
@angelofenoglio fixed and replied your comments

please take a look at [function set_backend_key](https://github.com/binbashar/leverage/blob/b7435b602f896acfb12a431a6e77f2383a80b3b0/leverage/container.py#L630)
I changed it so  it can generate the key when the S3 backend configuration exists and the --skip-validation was set

also, please look at [this line](https://github.com/binbashar/leverage/blob/b7435b602f896acfb12a431a6e77f2383a80b3b0/leverage/modules/terraform.py#L38)
I added a comment on how to call `--all-layers`... but maybe it has to be in a better place (aka documentation)

juanmatias commented:
@angelofenoglio 
regarding the scenarios for `get_location_type` https://github.com/binbashar/leverage/blob/99a8b039af9cc91d15806bfdcdac8d85475c194a/leverage/container.py#L313

I renamed the 'other under project' to 'sublayer'

it is used to run with `--layers` from inside a sublayer dir (aka not a layer nor an account dir, but still a valid one)
https://github.com/binbashar/leverage/blob/99a8b039af9cc91d15806bfdcdac8d85475c194a/leverage/modules/terraform.py#L196

angelofenoglio commented:
> please take a look at [function set_backend_key](https://github.com/binbashar/leverage/blob/b7435b602f896acfb12a431a6e77f2383a80b3b0/leverage/container.py#L630)
I changed it so it can generate the key when the S3 backend configuration exists and the --skip-validation was set

This looks good! And the small chart helps with understanding the logic.

> also, please look at [these lines](https://github.com/binbashar/leverage/blob/b7435b602f896acfb12a431a6e77f2383a80b3b0/leverage/modules/terraform.py#L38)
I added a comment on how to call --all-layers... but maybe it has to be in a better place (aka documentation)

The info in here works nicely as a quick reference but all the changes and different behaviors introduced in these last PRs should be clearly expressed in the documentation since they are going to be critical for the user.

> I renamed the 'other under project' to 'sublayer'

I think in this case is the other way around, so for example the dns layer which is on `shared/global/base-dns/binbash.com.ar`
`base-dns` contains no `.tf` files but holds the same place as the other `layers`, meanwhile `binbash.com.ar` is a subdirectory but is the actual directory where the `.tf` files are stored, a `sublayer` if you will.
Maybe the best solution is calling the one that holds the `.tf` files `layer` and any directory like `base-dns` something like `supralayer`? or `metalayer`? what do you think.
Opinions @diego-ojeda-binbash @exequielrafaela?

juanmatias commented:
@angelofenoglio value changed from `sublayer` to `layers-group`.

Changes made here:
https://github.com/binbashar/leverage/blob/53b6ba55416abd920bac2362c8fcb9dcae859075/leverage/container.py#L313
https://github.com/binbashar/leverage/blob/53b6ba55416abd920bac2362c8fcb9dcae859075/leverage/modules/terraform.py#L196
https://github.com/binbashar/leverage/blob/53b6ba55416abd920bac2362c8fcb9dcae859075/leverage/modules/aws.py#L182
